bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+#[derive(
+    Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
+)]
+pub struct ShareIdx(pub usize);
+
+/// Configuration of the FSM
+#[derive(Debug, Clone, Copy)]
+pub struct Config {
+    pub learn_timeout: Ticks,
+    pub rack_init_timeout: Ticks,
+    pub rack_secret_request_timeout: Ticks,
+}
+
+/// In memory state shared by all 4 FSM states
+#[derive(Debug)]
+pub struct FsmCommonData {
+    /// Unique IDs of this peer
+    pub id: Baseboard,
+
+    /// Configuration of the FSM
+    pub config: Config,
+
+    /// Unique IDs of connected peers
+    pub peers: BTreeSet<Baseboard>,
+
+    /// The current time in ticks since the creation of the FSM
+    pub clock: Ticks,
+
+    /// Any shares we are in the process of retrieving or a computed rack
+    /// secret. If no collection is ongoing and we are not holding a computed
+    /// rack secret, then the state is `Empty`. The state is always empty for
+    /// `State::Uninitialized` and `State::Learning`
+    pub rack_secret_state: RackSecretState,
+}
+
+impl FsmCommonData {
+    pub fn new(id: Baseboard, config: Config) -> FsmCommonData {
+        FsmCommonData {
+            id,
+            config,
+            peers: BTreeSet::new(),
+            clock: 0,
+            rack_secret_state: RackSecretState::Empty,
+        }
+    }
+
+    // Round-robin peer selection
+    pub fn next_peer(&self, current: &Baseboard) -> Option<Baseboard> {
+        if let Some(index) = self.peers.iter().position(|x| x == current) {
+            let next_index = (index + 1) % self.peers.len();
+            self.peers.iter().nth(next_index).cloned()
+        } else {
+            self.peers.first().cloned()
+        }
+    }
+
+    // Send a `GetShare` request to the newly connected peer if necessary
+    pub fn on_connect(&self, peer: Baseboard, rack_uuid: Uuid) -> Output {
+        if let RackSecretState::Retrieving { request_id, from, .. } =
+            &self.rack_secret_state
+        {
+            // We don't have a share from the peer and we weren't previously
+            // connected to the peer.
+            if !from.contains(&peer) && !self.peers.contains(&peer) {
+                let request = Request {
+                    id: *request_id,
+                    type_: RequestType::GetShare { rack_uuid },
+                };
+                return Output {
+                    persist: false,
+                    envelopes: vec![Envelope { to: peer, msg: request.into() }],
+                    api_output: None,
+                };
+            }
+        }
+        Output::none()
+    }
+}
+
+/// Metadata associated with a request that is keyed by `Baseboard`
+#[derive(Debug)]
+pub struct RequestMetadata {
+    pub request_id: Uuid,
+    pub expiry: Ticks,
+}
+
+#[derive(Zeroize, ZeroizeOnDrop)]
+pub struct Shares(Vec<Vec<u8>>);
+
+impl Shares {
+    fn insert(&mut self, share: Vec<u8>) {
+        self.0.push(share);
+    }
+}
+
+impl Debug for Shares {
```

Comment:
Is this manually implemented to avoid logging secrets?  If so, a comment would help.